cocotb/run.py

```diff
@@ -0,0 +1,175 @@
+"""A run class."""
+
+from subprocess import Popen
+import os
+import sys
+import sysconfig
+import cocotb
+import errno
+import distutils
+
+from setuptools import Extension
+from setuptools.command.build_ext import build_ext
+from setuptools.dist import Distribution
+
+cfg_vars = distutils.sysconfig.get_config_vars()
+for key, value in cfg_vars.items():
+    if type(value) == str:
+        cfg_vars[key] = value.replace("-Wstrict-prototypes", "")
+
+
+def _symlink_force(target, link_name):
+    try:
+        os.symlink(target, link_name)
+    except OSError as e:
+        if e.errno == errno.EEXIST:
+            os.remove(link_name)
+            os.symlink(target, link_name)
+        else:
+            raise e
+
+
+def _build_lib(lib, dist):
+    dist.ext_modules = [lib]
+    _build_ext = build_ext(dist)
+    _build_ext.finalize_options()
+
+    _build_ext.run()
+    out_lib = _build_ext.get_outputs()
+
+    lib_name = lib.name
+    lib_path = os.path.abspath(out_lib[0])
+    dir_name = os.path.dirname(lib_path)
+    ext_name = os.path.splitext(lib_path)[1][1:]
+
+    print("Building:", out_lib[0])
+
+    _symlink_force(
+        lib_path, os.path.join(os.path.abspath(dir_name), lib_name + "." + ext_name)
+    )
+
+    return dir_name
+
+
+def build_libs():
+    share_dir = os.path.join(os.path.dirname(cocotb.__file__), "share")
+
+    python_lib = sysconfig.get_config_var("LDLIBRARY")
+    python_lib_link = os.path.splitext(python_lib)[0][3:]
+
+    dist = Distribution()
+    dist.parse_config_files()
+
+    libcocotbutils = Extension(
+        "libcocotbutils",
+        include_dirs=[share_dir + "/include"],
+        sources=[share_dir + "/lib/utils/cocotb_utils.c"],
+    )
+
+    lib_path = _build_lib(libcocotbutils, dist)
```

Comment:
What happens if you set `dist.ext_modules = [libcocotbutils, libgpilog, libcocotb, libgpi]` and build them all at once?

Reply:
I do not know [yet\ how to st/set build directory so I take it after libcocotbutils one is compiled ;-(